File: 0127-word-ladder/0127-word-ladder.cpp
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        unordered_set<string> mp;
        for(int i = 0 ; i < wordList.size();i++){
            mp.insert(wordList[i]);
        }

        char a = 'a';
        queue<pair<string,int>> q;
        q.push({beginWord,1});

        while(!q.empty()){
            string word = q.front().first;
            int level = q.front().second;
            q.pop();

            if(word == endWord) return level;

            for(int i = 0; i < word.length() ; i++){
                for(int j = 0; j < 26;j++){
                    char og =word[i];
                    word[i] = char(a + j);
                    if(mp.find(word) != mp.end()){
                        q.push({word,level+1});
                        mp.erase(word);
                    }
                    word[i] = og;
                }
            }
        }

        return 0;
    }
};

File: 0127-word-ladder/0127-word-ladder.py
from collections import deque

class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        st = set() 
        for i in range(len(wordList)):
            st.add(wordList[i])
        
        q = deque()
        q.append((beginWord,1))

        while q:
            word, level = q.popleft()

            if(word==endWord):
                return level

            for i in range(len(word)):
                for j in range(26):
                    new_word = word[:i] + chr(ord('a') + j) + word[i+1:]
                    if new_word in st:
                        q.append((new_word,level+1))
                        st.remove(new_word)
        
        return 0

